proposals/call-tags/Overview.md

```diff
@@ -4,183 +4,254 @@
+When defining a `func` of type `[ti*] -> [to*]`, one can optionally specify `(call_tag $calltag*)`, where each `$calltag`'s type must be a supertype of with `[ti*] -> [to*]`.
+If so, the `funcref` returned by `func.ref` for this `func` handles exactly the call tags in `$calltag*`.
+
+`func_switch ($call_tag $func)* $func_switch` defines a new function, where the type of each `$call_tag` is a supertype of its corresponding `$func`'s type'.
+The new function has no type and cannot be directly called.
+However, the `funcref` returned by `func.ref` for this `func_switch` handles each `$call_tag` by calling the corresponding `$func`.
+If a call tag is not recognized, then if `$func_switch` is specified the call tag and arguments are forwarded to it, otherwise the fall-back handler of the call tag is tail-called with the arguments. 
+
+### Instructions
+
+`call_with_tag $call_tag : [funcref ti*] -> [to*]`, where `$call_tag : [ti*] -> [to*]`, calls the given `funcref` with the specified call tag using the values on the stack as the arguments.
+
+`call_return_with_tag $call_tag : [funcref ti*] -> [to*]`, where `$call_tag : [ti*] -> [to*]`, tail calls the given `funcref` with the specified call tag using the values on the stack as the arguments, where `[to*]` is a subtype of the result type of the function containing this instruction.
+
+## Implementation
+
+To implement `call_with_tag`, the arguments can be placed in registers and on the stack according to the calling convention prescribed by the call tag's signature type, except that the call tag's value takes the standard place of the first argument (which is moved to some remaining place).
+Then the code pointer in the `funcref` is called/jumped to.
+That code then switches on the value of the call tag, jumping-to/tail-calling the appropriate function while leaving all the arguments in their place except for moving the first argument.
+If the call tag is not recognized, then the code jumps-to/tail-calls the fall-back handler pointed to by the call tag, leaving all the arguments in their place but replacing the call-tag value with the value of the current `funcref`.
+
+[Transitioning from Structural to Nominal Code with Efficient Gradual Typing](https://www.cs.cornell.edu/~ross/publications/monnom/monnom-oopsla21.pdf) implemented call tags in LLVM by fixing an argument size used by all call tags.
+If a call tag's arity exceeded the argument size, then the remaining arguments were stored on the stack, and the last argument was instead a pointer to this stack-allocated tuple.
+It is unknown what the overhead of this approach to large-arity call tags is, because for their experiments no call tag's arity exceeded the fixed argument size.
+They did however measure the overhead of this LLVM-compilation approach for small-arity call tags compared to using typed function references, and they found that there was *no* overhead.
```

Comment:
To be clear, the typed function references in this experiment were in a non-Wasm context, right?